src/protocol/attribution/credit_capping.rs

```diff
@@ -0,0 +1,477 @@
+
+        [
+            Batch::try_from(shares[0].clone()).unwrap(),
+            Batch::try_from(shares[1].clone()).unwrap(),
+            Batch::try_from(shares[2].clone()).unwrap(),
+        ]
+    }
+
+    #[tokio::test]
+    pub async fn cap() {
+        // a result output from `AccumulateCredit`
+        const RAW_INPUT: &[[u128; 4]; 19] = &[
+            // [is_trigger, helper_bit, breakdown_key, credit]
+            [0, 0, 3, 0],
+            [0, 0, 4, 0],
+            [0, 1, 4, 19],
+            [1, 1, 0, 19],
+            [1, 1, 0, 9],
+            [1, 1, 0, 7],
+            [1, 1, 0, 6],
+            [1, 1, 0, 1],
+            [0, 0, 1, 0],
+            [1, 0, 0, 10],
+            [0, 0, 2, 15],
+            [1, 1, 0, 15],
+            [1, 1, 0, 12],
+            [0, 1, 2, 0],
+            [0, 1, 2, 10],
+            [1, 1, 0, 10],
+            [1, 1, 0, 4],
+            [0, 1, 5, 6],
+            [1, 1, 0, 6],
+        ];
+        const CAP: u32 = 20;
+        //
+        // Currently, this protocol doesn't produce outputs that we expect for
+        // some cases. One example is seen with this test case, where we have
+        // multiple source-trigger events sequences:
+        //
+        // Example:
+        // All events are from match key "1234".
+        // s = Source event, t = Trigger event
+        //
+        // s -> t -> t -> s -> s -> t -> t -> t -> ...
+        // ___________    _____________________
+        //   block 1             block 2
+        //
+        // A total contribution from this user "1243" should be capped at X,
+        // but the current logic caps the credits within each s-t "block",
+        //
+        // const EXPECTED: &[u128] = &[0, 0, 19, 0, 0, 0, 0, 0, 0, 0, 15, 0, 0, 0, 5, 0, 0, 0, 0];
+        const EXPECTED: &[u128] = &[0, 0, 19, 0, 0, 0, 0, 0, 0, 0, 20, 0, 0, 0, 10, 0, 0, 6, 0];
+
+        let world = TestWorld::<Fp32BitPrime>::new(QueryId);
+        let context = world.contexts();
+        let mut rng = StepRng::new(100, 1);
+
+        let shares = generate_shared_input(RAW_INPUT, &mut rng);
+
+        let [c0, c1, c2] = context;
+        let [s0, s1, s2] = shares;
+
+        let h0_future = CreditCapping::execute(c0, &s0, CAP);
+        let h1_future = CreditCapping::execute(c1, &s1, CAP);
+        let h2_future = CreditCapping::execute(c2, &s2, CAP);
+
+        let result = try_join!(h0_future, h1_future, h2_future).unwrap();
+
+        assert_eq!(result.0.len(), RAW_INPUT.len());
+        assert_eq!(result.1.len(), RAW_INPUT.len());
+        assert_eq!(result.2.len(), RAW_INPUT.len());
+        assert_eq!(result.0.len(), EXPECTED.len());
+
+        for (i, expected) in EXPECTED.iter().enumerate() {
+            let v = (
+                &result.0[i].credit,
+                &result.1[i].credit,
+                &result.2[i].credit,
+            )
+                .reconstruct();
+            assert_eq!(v.as_u128(), *expected);
+        }
+    }
```

Comment:
Can we please use the new test framework?

Reply:
I'll work on this in the following diff. Added TODO